langchain4j-core/src/main/java/dev/langchain4j/model/chat/request/ChatRequest.java

```diff
@@ -16,7 +17,8 @@
 public class ChatRequest {
 
     private final List<ChatMessage> messages;
-    private final List<ToolSpecification> toolSpecifications; // TODO add section with tools?
+    private final List<ToolSpecification> toolSpecifications;
+    // TODO tool mode
```

Comment:
Agreed, for the first iteration I would start with supporting 2 tool modes:
- LLM decides if it should call one or more tools: generic use case, used to add agentic behavior
- LLM is forced to call a (single) specific tool: can be used for structured data extraction with LLM providers that do not support structured output feature

[As fas as I see](https://docs.google.com/spreadsheets/d/1hOsbbAz9-qXCrnq_TXW-Q_AIAlssPLz9/edit?gid=532150960#gid=532150960), most LLM providers support those 2 modes. Other modes can be added later if needed

Reply:
Second scenario seems to be a specific case of `ANY` tool mode when `toolSpecifications` contain only a single item. So I guess we can start with this:

```
public enum ToolMode {
    AUTO, ANY
}
```

Reply:
So user could force a specific tool by:
```java
ChatRequest.builder()
    .toolSpecifications(List.of(weatherTool)) // single tool
    .toolMode(ANY)
    .build();
```

The downside of this approach is that user cannot specify multiple tools, but select only a specific one. I am not sure what is the use case for this (perharps provide LLM with more context about related tools?), but it seems that many LLM providers support this?

Reply:
I would be sad if all Gemini tool modes weren't supported by LangChain4j.

Currently, Gemini supports:

* **AUTO**: The default model behavior. The model decides to predict either a function call or a natural language response.
* **ANY**: The model is constrained to always predict a function call. If allowed_function_names is not provided, the model picks from all of the available function declarations. If allowed_function_names is provided, the model picks from the set of allowed functions.
* **NONE**: The model won't predict a function call. In this case, the model behavior is the same as if you don't pass any function declarations.

Reply:
> I would be sad if all Gemini tool modes weren't supported by LangChain4j.

We will try to support as many as possible, but **_for the first iteration_** I would focus on the most useful modes: `AUTO` and `ANY` (already mentioned above).
The only mode that I did not mention _for the first iteration_ is `NONE`. Is it even used by anyone?

Reply:
Yeah, I struggle to see when `NONE` would be used, but for the forced mode, I think one mandatory function call is a particular case of a mandatory call form a set of functions. So if we could from the first iteration support forced calls for one or more functions (instead of just one function). But OpenAI people would just pass one function and that would work, and Gemini people would be able to get multiple choice forced call to work just as well when several functions are passed.
I don't want we repeat a similar mistake with the API that was modeled after just OpenAI, when we can support more models from the beginning.

Reply:
Agreed, we can support `AUTO` (as a default mode) and `ANY` (one or more tools must be called from the provided list) from the beginning.

> I don't want we repeat a similar mistake with the API that was modeled after just OpenAI, when we can support more models from the beginning.

I do not want to "repeat a similar mistake" either, that's why I [analyze](https://docs.google.com/spreadsheets/d/1hOsbbAz9-qXCrnq_TXW-Q_AIAlssPLz9/edit?gid=532150960#gid=532150960) multiple LLM provider APIs.
But to be fair, the existing `ChatLanguageModel` API was "settled" one year ago before Devoxx and IIRC only OpenAI supported tools, so I gisagree that it was a mistake.